include/deal.II/cut/immersed_surface_quadrature.h

```diff
@@ -0,0 +1,152 @@
+   * surface integral over $\hat{S}$ according to
+   * @f[
+   * \int_{S} f(x) dS =
+   * \int_{S} f(x) |d\bar{S}| =
+   * \int_{\hat{S}} f(F_{K}(\hat{x})) \det(J) |\left( J^{-1} \right )^T d\hat{S}|.
+   * @f]
+   * This transformation is possible since the continuous surface elements
+   * are vectors: $d\bar{S}, d\hat{S} \in \mathbb{R}^{dim}$ which are parallel
+   * to the normals of $S$ and $\hat{S}$.
+   *
+   * By storing the quadrature points on the surface and a discrete surface
+   * element for each quadrature point, $\Delta \hat{S}_q$, the surface integral
+   * in real space can be approximated analogously
+   * @f[
+   * \int_{S} f(x) dS \approx
+   * \sum_{q} f(F_{K}(x_{q})) \det(J_q) |\left( J_q^{-1} \right)^T \Delta \hat{S}_q|.
+   * @f]
+   *
+   * @image html immersed_surface_quadrature.svg
+   *
+   * If the surface $\hat{S}$ is parameterized over a region $U\in \mathbb{R}^{dim-1}$
+   *
+   * @f[
+   *  \hat{S}=\{\bar{r}(\bar{u}) : \bar{u}\in U \}
+   * @f]
+   *
+   * the continuous surface element would take the form
+   * @f[
+   * d\hat{S}=\pm \left(
+   * \frac{\partial r_1}{\partial u_0},
+   * \frac{\partial r_0}{\partial u_0} \right)
+   * du_0
+   * @f]
+   * in 2D and
+   * @f[
+   * d\hat{S}=\pm \left(
+   * \frac{\partial (r_1,r_2)}{\partial (u_0,u_1)},
+   * \frac{\partial (r_2,r_0)}{\partial (u_0,u_1)},
+   * \frac{\partial (r_0,r_1)}{\partial (u_0,u_1)} \right)
+   * du_0 du_1
+   * @f]
+   * in 3D. Here, $\pm$ dependends on the surface orientation.
+   * Given that there is a quadrature for the region $U$ with points, $u_q$,
+   * and weights, $w_q$, the discrete surface element would be constructed in
+   * the same way by simply switching the area element against the quadrature
+   * weight
+   * @f[
+   * \Delta \hat{S}_q=\pm \left(
+   * \frac{\partial r_1}{\partial u_0},
+   * \frac{\partial r_0}{\partial u_0} \right)
+   * w_q
+   * @f]
+   * in 2D and
+   * @f[
+   * \Delta \hat{S}_q=\pm \left(
+   * \frac{\partial (r_1,r_2)}{\partial (u_0,u_1)},
+   * \frac{\partial (r_2,r_0)}{\partial (u_0,u_1)},
+   * \frac{\partial (r_0,r_1)}{\partial (u_0,u_1)} \right)
+   * w_q
+   * @f]
+   * in 3D.
+   *
+   * @image html immersed_surface_parameterization.svg
+   */
+  template<int dim>
+  class ImmersedSurfaceQuadrature : public Subscriptor
+  {
+  public:
+
+    /**
+     * Default constructor initialize the quadrature with no quadrature points.
+     */
+    ImmersedSurfaceQuadrature();
+
+    /**
+     * Extend the given formula by an additional quadrature @p point.
+     * The @p surface_element is a normal to the surface which is scaled so
+     * that its norm equals the quadrature weight, see class description above.
+     *
+     * @note This function should only be used during construction of the
+     * quadrature formula.
+     */
+    void add_point(const Point<dim> &point,
+                   const Tensor<1, dim> &surface_element);
```

Comment:
Could you explain why you want to add points one by one rather than simply making a constructor that takes a list of points? I guess this is due to the use case, but since it is unusual to be able to add points on the fly I think it would be good to have some background.